### ISSUE TITLE
Allow arbitrary URL manipulation dynamically for RobotsTxt even in the backend

### ISSUE THREAD

**Issue by mario-naether** [mblaschke/TYPO3-metaseo#358]
**MetaSEO version**: 2.0.3

**TYPO3 version**: 7.6.11

**PHP version**: 5.6.5

RealUrl version (optional): 2.1.8

Ich hab folgende TypoScript Konfiguration hinterlegt.
```
plugin.metaseo.robotsTxt.default{
    20 = TEXT
    20.value (
Disallow: /*?id=*
Disallow: /*&type=103
Disallow: /*&sword_list*
Disallow: /*&no_cache*
  )
    90.noTrimWrap = |Sitemap: ||
    90.typolink.returnLast = url
    90.typolink.parameter = 261
    90.typolink.additionalParams  = &type=841132
  }
```

und bekomme dann ein PHP Fatal Error wenn ich im Backend über das Kontrollzentrum die SEO Konfiguration bearbeiten möchte.

`
```
Fatal error: Call to a member function getPage_noCheck() on string in ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php on line 6362

Call Stack:
    0.0000     237944   1. {main}() ***typo3/index.php:0
    0.0000     238352   2. call_user_func:{***typo3/index.php:21}() ***typo3/index.php:21
    0.0000     238392   3. {closure:***typo3/index.php:18-21}() ***typo3/index.php:21
    0.0608    5258360   4. TYPO3\CMS\Backend\Http\Application->run() ***typo3/index.php:20
    0.0655    5384840   5. TYPO3\CMS\Core\Core\Bootstrap->handleRequest() ***typo3/sysext/backend/Classes/Http/Application.php:94
    0.0669    5407648   6. TYPO3\CMS\Backend\Http\RequestHandler->handleRequest() ***typo3/sysext/core/Classes/Core/Bootstrap.php:302
    0.4404   12154520   7. TYPO3\CMS\Backend\Http\RequestHandler->dispatch() ***typo3/sysext/backend/Classes/Http/RequestHandler.php:94
    0.4417   12192072   8. TYPO3\CMS\Backend\Http\RouteDispatcher->dispatch() ***typo3/sysext/backend/Classes/Http/RequestHandler.php:161
    0.4729   12983496   9. call_user_func_array:{***typo3/sysext/backend/Classes/Http/RouteDispatcher.php:54}() ***typo3/sysext/backend/Classes/Http/RouteDispatcher.php:54
    0.4729   12983704  10. TYPO3\CMS\Backend\Controller\EditDocumentController->mainAction() ***typo3/sysext/backend/Classes/Http/RouteDispatcher.php:54
    0.4754   13010232  11. TYPO3\CMS\Backend\Controller\EditDocumentController->main() ***typo3/sysext/backend/Classes/Controller/EditDocumentController.php:2018
    0.4759   13020064  12. TYPO3\CMS\Backend\Controller\EditDocumentController->makeEditForm() ***typo3/sysext/backend/Classes/Controller/EditDocumentController.php:1000
    0.6214   14084664  13. TYPO3\CMS\Backend\Form\Container\OuterWrapContainer->render() ***typo3/sysext/backend/Classes/Controller/EditDocumentController.php:1192
    0.6218   14096176  14. TYPO3\CMS\Backend\Form\Container\FullRecordContainer->render() ***typo3/sysext/backend/Classes/Form/Container/OuterWrapContainer.php:53
    0.6267   14111288  15. TYPO3\CMS\Backend\Form\Container\TabsContainer->render() ***typo3/sysext/backend/Classes/Form/Container/FullRecordContainer.php:84
    0.6340   14195896  16. TYPO3\CMS\Backend\Form\Container\PaletteAndSingleContainer->render() ***typo3/sysext/backend/Classes/Form/Container/TabsContainer.php:86
    0.6344   14208424  17. TYPO3\CMS\Backend\Form\Container\PaletteAndSingleContainer->createPaletteContentArray() ***typo3/sysext/backend/Classes/Form/Container/PaletteAndSingleContainer.php:117
    0.6362   14236944  18. TYPO3\CMS\Backend\Form\Container\SingleFieldContainer->render() ***typo3/sysext/backend/Classes/Form/Container/PaletteAndSingleContainer.php:215
    0.6368   14255440  19. TYPO3\CMS\Backend\Form\Element\UserElement->render() ***typo3/sysext/backend/Classes/Form/Container/SingleFieldContainer.php:186
    0.6368   14259136  20. TYPO3\CMS\Core\Utility\GeneralUtility::callUserFunction() ***typo3/sysext/backend/Classes/Form/Element/UserElement.php:43
    0.6371   14264344  21. call_user_func_array:{***typo3/sysext/core/Classes/Utility/GeneralUtility.php:4377}() ***typo3/sysext/core/Classes/Utility/GeneralUtility.php:4377
    0.6371   14265872  22. Metaseo\Metaseo\Hook\TCA\RobotsTxtDefault->main() ***typo3/sysext/core/Classes/Utility/GeneralUtility.php:4377
    1.1556   16422496  23. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->cObjGetSingle() typo3conf/ext/metaseo/Classes/Hook/TCA/RobotsTxtDefault.php:122
    1.1566   16430936  24. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->render() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:859
    1.1566   16431192  25. TYPO3\CMS\Frontend\ContentObject\ContentObjectArrayContentObject->render() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:943
    1.1566   16431344  26. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->cObjGet() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectArrayContentObject.php:41
    1.1570   16436312  27. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->cObjGetSingle() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:805
    1.1571   16436616  28. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->render() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:859
    1.1571   16436712  29. TYPO3\CMS\Frontend\ContentObject\TextContentObject->render() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:943
    1.1571   16437288  30. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->stdWrap() ***typo3/sysext/frontend/Classes/ContentObject/TextContentObject.php:43
    1.1571   16439408  31. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->stdWrap_typolink() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:2463
    1.1571   16440856  32. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->typoLink() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:3528
    1.1590   16446936  33. TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer->detectLinkTypeFromLinkParameter() ***typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php:6469
```

`

**Comment by thomaszbz** [comment]
You're trying to use TYPO3's URL manipulation functionality for the frontend in the backend:

`90.typolink.returnLast = url`

I guess this overstretches some of the infrastructure TYPO3 provides.

Eventually, it also overstretches the way this variable of metase once was intended to be used: (presumably) statically. I'm not sure on the latter, would need to look it up.

Btw, what exactly is your typoscript code supposed to do? Maybe there is a better solution. Somehow looks tweaked to me.

**Comment by mario-naether** [comment]
Hi,

you are using this `Metaseo\Metaseo\Page\RobotsTxtPage` in your Backend Module to view the robotos.txt output.
This RobotsTxtPage class use some FE objects like `$GLOBALS['TSFE']->cObj`, this work good untill the typo link function was called.
The link function error occured because  this object `$GLOBALS['TSFE']->sys_page` isn't initialized.

If you call `$GLOBALS['TSFE']->fetch_the_id();` in RobotsTxtPage->main(), the"->sys_page" object should be initialized and all the typolink stuff should be work!?

mfg mario

**Comment by thomaszbz** [comment]
@mario-naether Do you think this might be of use for others? Do you want to provide a patch?

**Comment by mario-naether** [comment]
I have create a pull request.
https://github.com/mblaschke/TYPO3-metaseo/pull/360

is it ok for you?

**Comment by thomaszbz** [comment]
Reproduced the exception against current 2.1.0-dev on develop (ea646bab72fe2c80397e1d9b097bcce0202dd17f):

When clicking on the root page

![auswahl_252](https://cloud.githubusercontent.com/assets/11200727/25168697/d0a8a1a2-24e4-11e7-9d01-85d8697d7886.png)

then I get the following exception with the TypoScript code you provided:

> Fatal error: Call to a member function getPage_noCheck() on string in /var/webadmin/typo3_src/typo3/sysext/frontend/Classes/ContentObject/ContentObjectRenderer.php on line 6362

I think we could also classify that as a bug.

Your fix fixes it, just that I backported it to PHP 5.3, added a missing namespace and cherry-picked it for 2.1.0
